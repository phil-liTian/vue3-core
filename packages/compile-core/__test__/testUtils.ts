import { isArray, PatchFlagNames, PatchFlags, ShapeFlags } from '../../shared'
import {
  ElementNode,
  locStub,
  NameSpaces,
  NodeTypes,
  VNodeCall,
} from '../src/ast'

export function createElementWithCodegen(
  tag: VNodeCall['tag'],
  props: VNodeCall['props'],
  children?: VNodeCall['children'],
  patchFlag?: VNodeCall['patchFlag'],
): ElementNode {
  return {
    type: NodeTypes.ELEMENT,
    tag: 'div',
    props,
    children,
    loc: locStub,
    ns: NameSpaces.HTML,
    codegenNode: {
      type: NodeTypes.VNODE_CALL,
      patchFlag,
      tag,
      props,
      children,
    },
  }
}

type Flags = PatchFlags | ShapeFlags
export function genFlagText(
  flag: Flags | Flags[],
  names: { [k: number]: string } = PatchFlagNames,
): string {
  if (isArray(flag)) {
    let f = 0
    flag.forEach(ff => {
      f |= ff
    })
    return `${f} /* ${flag.map(f => names[f]).join(', ')} */`
  } else {
    return `${flag} /* ${names[flag]} */`
  }
}
