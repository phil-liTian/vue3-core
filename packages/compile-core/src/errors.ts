export enum ErrorCodes {
  X_INVALID_END_TAG,
}
