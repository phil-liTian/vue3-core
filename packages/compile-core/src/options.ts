import { NodeTransform } from './transform'

// 定义transtorm 的选项
export interface TransformOptions {
  nodeTransforms?: NodeTransform[]


  prefixIdentifiers?: boolean
}

// 定义generate 的选项
export interface CodegenOptions {
  /**
   * module模式 使用es模块化语法导出helpers
   * @default function
   */
  mode?: 'module' | 'function'

  /**
   * 是否添加 `with (this) { ... }`
   * mode为function是 默认不添加
   * @default mode === 'module'
   */
  prefixIdentifiers?: boolean

  /**
   * 可自定义导入helper的地方
   * @default vue
   */
  runtimeModuleName?: string

  /**
   * 可自定义导入全局变量的地方
   * @default Vue
   */
  runtimeGlobalName?: string

  /**
   * 是否优化导入
   * (用作webpack 代码分割)
   * @default false
   */
  optimizeImports?: boolean
}
