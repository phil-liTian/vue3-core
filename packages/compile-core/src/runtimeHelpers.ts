export const CREATE_VNODE: unique symbol = Symbol('createVNode')
export const RESOLVE_DIRECTIVE: unique symbol = Symbol('resolveDirective')

export const helperNameMap = {
  [CREATE_VNODE]: 'createVNode',
  [RESOLVE_DIRECTIVE]: 'resolveDirective',
}
