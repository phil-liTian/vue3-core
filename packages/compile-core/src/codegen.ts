import { isString } from '../../shared'
import {
  CallExpression,
  CommentNode,
  CompoundExpressionNode,
  ConditionalExpression,
  ExpressionNode,
  InterpolationNode,
  NodeTypes,
  ObjectExpression,
  RootNode,
  SimpleExpressionNode,
  VNodeCall,
} from './ast'
import { CodegenOptions } from './options'
import {
  CREATE_COMMENT,
  helperNameMap,
  OPEN_BLOCK,
  RESOLVE_COMPONENT,
  RESOLVE_DIRECTIVE,
  TO_DISPLAY_STRING,
} from './runtimeHelpers'
import { isSimpleIdentifier, isValidAssetId } from './utils'

export enum NewlineType {
  Start = 0,
  End = -1,
  None = -2,
  Unknown = -3,
}

export interface CodegenContext extends CodegenOptions {
  code: string
  indentLevel: number // 缩进的个数
  helper: (key: symbol) => void
  push: (code: string, newLineIndex?: number) => void
  newline: () => void
  indent: () => void
  deindent: (withoutNewline?: boolean) => void // 回退缩进
}

const aliasHelper = (s: symbol) => `${helperNameMap[s]}: _${helperNameMap[s]}`

// 创建ast的上下文
function createCodegenContext(
  ast,
  {
    mode = 'function',
    runtimeModuleName = 'vue',
    runtimeGlobalName = 'Vue',
    optimizeImports = false,
    prefixIdentifiers = mode === 'module',
  }: CodegenOptions,
): CodegenContext {
  const context: CodegenContext = {
    code: '',
    mode,
    optimizeImports,
    runtimeModuleName,
    runtimeGlobalName,
    prefixIdentifiers,
    indentLevel: 0,

    helper(key) {
      return `_${helperNameMap[key]}`
    },
    push(code, newLineIndex = NewlineType.None) {
      context.code += code
    },

    indent() {
      newline(++context.indentLevel)
    },

    deindent(withoutNewline = false) {
      if (withoutNewline) {
        --context.indentLevel
      } else {
        newline(--context.indentLevel)
      }
    },

    // 可实现待缩减换行
    newline() {
      newline(context.indentLevel)
    },
  }

  function newline(n: number) {
    context.push('\n' + '  '.repeat(n))
  }

  return context
}

// 主方法
export function generate(ast: RootNode, options: CodegenOptions = {}) {
  const context = createCodegenContext(ast, options)

  const { push, mode, prefixIdentifiers, newline } = context
  const { temps, components, directives } = ast

  const helpers = Array.from(ast.helpers!)
  const hasHelper = helpers.length > 0
  const useWithBlock = !prefixIdentifiers && mode !== 'module'

  // 生成函数前导码
  if (mode === 'module') {
    genModulePreamble(ast, context)
  } else {
    genFunctionPreamble(ast, context)
  }
  const functionName = 'render'
  const args = ['_ctx', '_cache']
  const signarture = args.join(', ')
  push(`function ${functionName}(${signarture}) {`)

  if (useWithBlock) {
    push(`with (_ctx)`)

    if (hasHelper) {
      push(`const { ${helpers.map(aliasHelper).join(', ')} } = _Vue\n`)
    }
  }

  // 处理components
  if (components?.length) {
    genAssets(components, 'component', context)
    if (temps > 0) {
      newline()
    }
  }

  // 处理directives
  if (directives?.length) {
    genAssets(directives, 'directive', context)
    if (temps > 0) {
      newline()
    }
  }

  if (temps > 0) {
    push('let')
    for (let i = 0; i < temps; i++) {
      push(`${i > 0 ? ',' : ''} _temp${i}`)
    }
  }

  push('\n')
  push(`return `)

  // 如何生成的codegenNode??? transform??
  if (ast.codegenNode) {
    genNode(ast.codegenNode, context)
  }

  push(`}`)

  return {
    code: context.code,
  }
}

// 指定mode为module时, 使用es模块化导入
function genModulePreamble(ast: RootNode, context: CodegenContext) {
  const { push, runtimeModuleName, optimizeImports } = context
  const { helpers } = ast

  if (helpers?.size) {
    const helpers = Array.from(ast.helpers!)

    if (optimizeImports) {
      push(
        `import { ${helpers.map(h => `${helperNameMap[h]}`).join(', ')} } from ${JSON.stringify(runtimeModuleName)}\n`,
      )

      push(
        `\n//Binding optimization for webpack code-split\nconst ${helpers.map(h => `_${helperNameMap[h]} = ${helperNameMap[h]}`).join(', ')}`,
      )
    } else {
      push(
        `import { ${helpers.map(h => `${helperNameMap[h]} as _${helperNameMap[h]}`).join(', ')} } from ${JSON.stringify(runtimeModuleName)}\n`,
      )
    }
  }
}

// 函数前导
function genFunctionPreamble(ast: RootNode, context) {
  const {
    push,
    runtimeModuleName,
    mode,
    runtimeGlobalName,
    prefixIdentifiers,
  } = context
  const {} = ast

  const helper = Array.from(ast.helpers!)
  if (helper.length) {
    if (prefixIdentifiers) {
      // 将导入语句放到整个函数块之前
      push(`const { ${helper.map(aliasHelper).join(', ')} } = Vue\n`)
    } else {
      push(`const _Vue = ${runtimeGlobalName}\n`)
    }
  }
  genHoist(ast.hoists, context)

  push('return ')
}

// 处理 components、directives
function genAssets(
  assets: string[],
  type: 'component' | 'directive',
  { push, helper, newline }: CodegenContext,
) {
  const resolver = helper(
    type === 'component' ? RESOLVE_COMPONENT : RESOLVE_DIRECTIVE,
  )
  for (let i = 0; i < assets.length; i++) {
    let id = assets[i]

    const maybeSetReference = id.endsWith('__self')
    if (maybeSetReference) {
      id = id.slice(0, -6)
    }

    push(
      `const ${isValidAssetId(id, type)} = ${resolver}(${JSON.stringify(id)}${maybeSetReference ? ', true' : ''})`,
    )

    // 不是最后一个assets则末尾换行处理
    if (i < assets.length - 1) {
      newline()
    }
  }
}

// 处理静态提升
function genHoist(hoist, context) {
  const { push, newline } = context
  if (!hoist?.length) return
  for (let i = 0; i < hoist.length; i++) {
    const exp = hoist[i]
    if (exp) {
      push(`const _hoisted_${i + 1} = `)
      genNode(exp, context)

      newline()
    }
  }
}

// 处理node的核心入口函数
function genNode(node: any, context) {
  switch (node.type) {
    case NodeTypes.ELEMENT:
    case NodeTypes.IF:
    case NodeTypes.FOR:
      genNode(node.codegenNode, context)
      break
    case NodeTypes.TEXT:
      genText(node, context)
      break

    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break
    case NodeTypes.JS_OBJECT_EXPRESSION: {
      genObjectExpression(node, context)
      break
    }
    case NodeTypes.INTERPOLATION: {
      genInterpolation(node, context)
      break
    }
    case NodeTypes.COMMENT: {
      genComment(node, context)
      break
    }
    case NodeTypes.COMPOUND_EXPRESSION: {
      genCompoundExpression(node, context)
      break
    }
    case NodeTypes.JS_CONDITIONAL_EXPRESSION: {
      genConditionalExpression(node, context)
      break
    }
    case NodeTypes.VNODE_CALL: {
      genVNodeCall(node, context)
      break
    }
    case NodeTypes.JS_CALL_EXPRESSION: {
      genCallExpression(node, context)
      break
    }
  }
}

// 处理纯文本
function genText(node, context) {
  const { push } = context
  push(JSON.stringify(node.content))
}

// 处理简单表达式
function genExpression(node: SimpleExpressionNode, context: CodegenContext) {
  const { push } = context
  const { isStatic, content } = node
  push(isStatic ? JSON.stringify(content) : content)
}

// 处理js对象
function genObjectExpression(node: ObjectExpression, context: CodegenContext) {
  const { properties } = node
  const { push, newline } = context
  if (!properties.length) {
    return
  }

  push('{ ')

  for (let i = 0; i < properties.length; i++) {
    const { key, value } = properties[i]
    // 处理对象的key
    genExpressionAsPropertyKey(key, context)
    push(': ')
    // 处理对象的value
    genNode(value, context)
    // 同样的 如果不是第一个 就换行展示
    if (i < properties.length - 1) {
      newline()
    }
  }

  push(' }')
}

// 处理插值
function genInterpolation(node: InterpolationNode, context: CodegenContext) {
  const { content } = node
  const { push, helper } = context
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(content, context)
  push(')')
}

// 处理注释
function genComment(node: CommentNode, context: CodegenContext) {
  const { push, helper } = context
  push(`${helper(CREATE_COMMENT)}(${JSON.stringify(node.content)})`)
}

// 处理CompoundExpressionNode的node, children是一个数组, 是各种元素的集合
function genCompoundExpression(
  node: CompoundExpressionNode,
  context: CodegenContext,
) {
  const { push } = context
  for (let i = 0; i < node.children.length; i++) {
    const child = node.children[i]
    if (isString(child)) {
      push(child as string)
    } else {
      genNode(child, context)
    }
  }
}

// 处理条件语句
function genConditionalExpression(
  node: ConditionalExpression,
  context: CodegenContext,
) {
  const { push, indent, newline } = context
  const { test, consequent, alternate, newline: needNewline } = node
  if (test.type === NodeTypes.SIMPLE_EXPRESSION) {
    genExpression(test, context)
  }
  needNewline && indent()
  needNewline || push(` `)
  push('? ')
  genNode(consequent, context)
  context.indentLevel--
  needNewline && newline()
  push(': ')
  genNode(alternate, context)
}

function genVNodeCall(node: VNodeCall, context: CodegenContext) {
  const { isBlock, disableTracking } = node
  const { push, helper } = context
  if (isBlock) {
    push(`(${helper(OPEN_BLOCK)}(${disableTracking ? `true` : ''})`)
  }
}

function genCallExpression(node: CallExpression, context: CodegenContext) {
  const { push, helper } = context
}

// ---------------------------- utils -------------------------------
// 处理js对象的key
function genExpressionAsPropertyKey(
  node: ExpressionNode,
  context: CodegenContext,
) {
  const { push } = context
  if (node.isStatic) {
    const text = isSimpleIdentifier(node.content)
      ? node.content
      : JSON.stringify(node.content)

    push(text)
  }
}
