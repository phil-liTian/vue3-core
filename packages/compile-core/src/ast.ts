import { isString, PatchFlags } from '../../shared'
import { CREATE_ELEMENT_VNODE } from './runtimeHelpers'

// ------------------------------types------------------------------------------
export enum NameSpaces {
  HTML,
}

export enum NodeTypes {
  TEXT,
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  COMMENT,
  ELEMENT,
  ROOT,
  COMPOUND_EXPRESSION,

  // containers
  IF,
  FOR,

  // codegen
  JS_CONDITIONAL_EXPRESSION,
  VNODE_CALL,
  JS_PROPERTY,
  JS_OBJECT_EXPRESSION,
  JS_ARRAY_EXPRESSION,
  JS_FUNCTION_EXPRESSION,
  JS_CALL_EXPRESSION,
}

export enum ElementTypes {
  ELEMENT,
}

export type propsExpression = ObjectExpression

export type ElementNode = any

export type TemplateChildNode =
  | ElementNode
  | TextNode
  | CommentNode
  | ForNode
  | InterpolationNode
  | CompoundExpressionNode

// loc记录行、列、偏移量
export interface Position {
  line: number
  offset: number
  column: number
}

export interface SourceLocation {
  start: Position
  end: Position
  source: string
}

// -------------------------Node Types-------------------------------------------------
export interface Node {
  loc: SourceLocation
  type: NodeTypes
}

export interface TextNode extends Node {
  type: NodeTypes.TEXT
  content: string
}

export type JSChildNode =
  | ConditionalExpression
  | ExpressionNode
  | ObjectExpression

export interface CallExpression extends Node {
  type: NodeTypes.JS_CALL_EXPRESSION
  callee: string | symbol
  arguments: (string | symbol)[]
}

export interface RootNode extends Node {
  type: NodeTypes.ROOT
  source: string
  children: Node[]
  codegenNode?: TemplateChildNode
  helpers?: Set<symbol>
  temps: number
  directives?: string[]
  components?: string[]
  hoists?: (JSChildNode | null)[]
}

export enum ConstantTypes {
  NOT_CONSTANT = 0,
  CAN_SKIP_PATCH,
  CAN_CACHE,
  CAN_STRINGIFY,
}

export interface SimpleExpressionNode extends Node {
  type: NodeTypes.SIMPLE_EXPRESSION
  content: string
  isStatic: boolean
  constType: ConstantTypes
}

export type ExpressionNode = SimpleExpressionNode

export interface InterpolationNode extends Node {
  type: NodeTypes.INTERPOLATION
  content: ExpressionNode | string
}

export interface CommentNode extends Node {
  type: NodeTypes.COMMENT
  content: string
}

export interface ConditionalExpression extends Node {
  type: NodeTypes.JS_CONDITIONAL_EXPRESSION
  test: JSChildNode
  consequent: JSChildNode
  alternate: JSChildNode
  newline: boolean
}

export interface IfConditionalExpression extends ConditionalExpression {
  consequent: any
  alternate: any
}

export interface ForNode extends Node {
  type: NodeTypes.FOR
  source: ExpressionNode
  valueAlias: ExpressionNode | undefined
  keyAlias: ExpressionNode | undefined
  objectIndexAlias: ExpressionNode | undefined
  parseResult: ForParseResult
  children: TemplateChildNode[]
}

export interface ForParseResult {
  source: ExpressionNode
  key: ExpressionNode | undefined
  value: ExpressionNode | undefined
  index: ExpressionNode | undefined
  finalized: boolean
}

export interface VNodeCall extends Node {
  tag: string | symbol | CallExpression
  props: propsExpression | undefined
  children: TemplateChildNode[] | string | undefined
  type: NodeTypes.VNODE_CALL

  dynamicProps: undefined | string

  isBlock: boolean
  isComponent: boolean
  disableTracking: boolean
  patchFlag: undefined | PatchFlags
}

export interface CompoundExpressionNode extends Node {
  type: NodeTypes.COMPOUND_EXPRESSION

  children: (
    | TextNode
    | InterpolationNode
    | SimpleExpressionNode
    | CompoundExpressionNode
    | string
  )[]
}

export interface Property extends Node {
  type: NodeTypes.JS_PROPERTY
  key: ExpressionNode
  value: JSChildNode
}

export interface ObjectExpression extends Node {
  type: NodeTypes.JS_OBJECT_EXPRESSION
  properties: Array<Property>
}

export type ParentNode = ElementNode | RootNode

// AST Utilities ---------------------------------------------------------------

export const locStub: SourceLocation = {
  start: { line: 1, column: 1, offset: 0 },
  end: { line: 1, column: 1, offset: 0 },
  source: '',
}

export function createRoot(children, source): RootNode {
  return {
    type: NodeTypes.ROOT,
    source,
    children,
    loc: locStub,
    codegenNode: undefined,
    temps: 0,
  }
}

// 创建一个简单的表达式
export function createSimpleExpression(
  content: SimpleExpressionNode['content'],
  isStatic: SimpleExpressionNode['isStatic'] = false,
  loc: SourceLocation = locStub,
  constType: ConstantTypes = ConstantTypes.NOT_CONSTANT,
): SimpleExpressionNode {
  return {
    type: NodeTypes.SIMPLE_EXPRESSION,
    content,
    isStatic,
    loc,
    constType,
  }
}

// 创建一个对象属性
export function createObjectProperty(
  key: Property['key'],
  value: Property['value'],
): Property {
  return {
    type: NodeTypes.JS_PROPERTY,
    key,
    value,
    loc: locStub,
  }
}

// 创建一个对象表达式
export function createObjectExpression(
  properties: ObjectExpression['properties'],
  loc: SourceLocation = locStub,
): ObjectExpression {
  return {
    type: NodeTypes.JS_OBJECT_EXPRESSION,
    loc,
    properties,
  }
}

// 创建一个插值表达式
export function createInterpolation(
  content: InterpolationNode['content'] | string,
  loc: SourceLocation,
): InterpolationNode {
  return {
    type: NodeTypes.INTERPOLATION,
    content: isString(content)
      ? createSimpleExpression(content as string, false, loc)
      : content,
    loc: locStub,
  }
}

// 创建一个复合类型 及 element 内 既有text又有interpolation的表达式 会组成一个复合类型
export function createCompoundExpression(
  children: CompoundExpressionNode['children'],
  loc: SourceLocation = locStub,
): CompoundExpressionNode {
  return {
    type: NodeTypes.COMPOUND_EXPRESSION,
    loc,
    children,
  }
}

// 创建一个条件表达式
export function createConditionalExpression(
  test: ConditionalExpression['test'],
  consequent: ConditionalExpression['consequent'],
  alternate: ConditionalExpression['alternate'],
  newline = true,
): ConditionalExpression {
  return {
    type: NodeTypes.JS_CONDITIONAL_EXPRESSION,
    test,
    consequent,
    alternate,
    newline,
    loc: locStub,
  }
}

// 创建一个call表达式
export function createCallExpression<T extends CallExpression['callee']>(
  callee: T,
  args: CallExpression['arguments'] = [],
  loc: SourceLocation = locStub,
) {
  return {
    type: NodeTypes.JS_CALL_EXPRESSION,
    callee,
    arguments: args,
    loc,
  }
}

// 创建vnode
export function createVNodeCall(
  context,
  tag: VNodeCall['tag'],
  props: VNodeCall['props'],
  children: VNodeCall['children'],
) {
  return {
    type: NodeTypes.VNODE_CALL,
    props,
    tag,
    children,
  }
}

export function getVNodeHelper() {
  return CREATE_ELEMENT_VNODE
}
