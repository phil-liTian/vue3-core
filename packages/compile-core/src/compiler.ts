import { generate } from './codegen'
import { baseParse } from './parser'
import { transform } from './transform'

export function baseCompiler(source) {
  const ast = baseParse(source)

  transform(ast)

  return generate(ast)
}
