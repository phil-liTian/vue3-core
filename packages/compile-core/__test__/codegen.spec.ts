import { describe, expect, test } from 'vitest'
import { generate } from '../src/codegen'
import {
  createConditionalExpression,
  createCallExpression,
  createCompoundExpression,
  createInterpolation,
  createObjectExpression,
  createObjectProperty,
  createSimpleExpression,
  locStub,
  NodeTypes,
  RootNode,
} from '../src/ast'

import {
  CREATE_COMMENT,
  CREATE_VNODE,
  FRAGMENT,
  helperNameMap,
  RENDER_LIST,
  RESOLVE_COMPONENT,
  RESOLVE_DIRECTIVE,
  TO_DISPLAY_STRING,
  CREATE_ELEMENT_VNODE,
} from '../src/runtimeHelpers'
import { createElementWithCodegen, genFlagText } from './testUtils'
import { PatchFlags } from '../../shared'

function createRoot(options: Partial<RootNode> = {}): RootNode {
  return {
    type: NodeTypes.ROOT,
    source: '',
    children: [],
    helpers: new Set(),
    loc: locStub,
    temps: 0,
    ...options,
  } as RootNode
}

describe('compiler: codegen', () => {
  test('mode preamble', () => {
    const root = createRoot({
      helpers: new Set([CREATE_VNODE, RESOLVE_DIRECTIVE]),
    })

    const { code } = generate(root, { mode: 'module' })

    expect(code).toMatch(
      `import { ${helperNameMap[CREATE_VNODE]} as _${helperNameMap[CREATE_VNODE]}, ${helperNameMap[RESOLVE_DIRECTIVE]} as _${helperNameMap[RESOLVE_DIRECTIVE]} } from "vue"`,
    )
    expect(code).toMatchSnapshot()
  })

  test('w/ optimizeImports: true', () => {
    const root = createRoot({
      helpers: new Set([CREATE_VNODE, RESOLVE_DIRECTIVE]),
    })

    const { code } = generate(root, { mode: 'module', optimizeImports: true })

    expect(code).toMatch(
      `import { ${helperNameMap[CREATE_VNODE]}, ${helperNameMap[RESOLVE_DIRECTIVE]} } from "vue"`,
    )

    expect(code).toMatch(
      `const _${helperNameMap[CREATE_VNODE]} = ${helperNameMap[CREATE_VNODE]}, _${helperNameMap[RESOLVE_DIRECTIVE]} = ${helperNameMap[RESOLVE_DIRECTIVE]}`,
    )
  })

  test('function mode preamble', () => {
    const root = createRoot({
      helpers: new Set([CREATE_VNODE, RESOLVE_DIRECTIVE]),
    })
    const { code } = generate(root, { mode: 'function' })
    expect(code).toMatch(`const _Vue = Vue`)

    expect(code).toMatch(
      `const { ${helperNameMap[CREATE_VNODE]}: _${helperNameMap[CREATE_VNODE]}, ${helperNameMap[RESOLVE_DIRECTIVE]}: _${helperNameMap[RESOLVE_DIRECTIVE]} } = _Vue`,
    )
    expect(code).toMatchSnapshot()
  })

  test('function mode preamble w/ prefixIdentifiers: true', () => {
    const root = createRoot({
      helpers: new Set([CREATE_VNODE, RESOLVE_DIRECTIVE]),
    })
    const { code } = generate(root, {
      mode: 'function',
      prefixIdentifiers: true,
    })
    expect(code).not.toMatch(`const _Vue = Vue`)
    expect(code).toMatch(
      `const { ${helperNameMap[CREATE_VNODE]}: _${helperNameMap[CREATE_VNODE]}, ${helperNameMap[RESOLVE_DIRECTIVE]}: _${helperNameMap[RESOLVE_DIRECTIVE]} } = Vue`,
    )
    expect(code).toMatchSnapshot()
  })

  test('assets + temps', () => {
    const root = createRoot({
      components: [`Foo`, `bar-baz`, `barbaz`, `Qux__self`],
      directives: ['my_dir_0', 'my_dir_1'],
      temps: 3,
    })

    const { code } = generate(root, { mode: 'function' })
    expect(code).toMatch(
      `const _component_Foo = _${helperNameMap[RESOLVE_COMPONENT]}("Foo")\n`,
    )

    expect(code).toMatch(
      `const _component_bar_baz = _${helperNameMap[RESOLVE_COMPONENT]}("bar-baz")\n`,
    )

    expect(code).toMatch(
      `const _component_barbaz = _${helperNameMap[RESOLVE_COMPONENT]}("barbaz")\n`,
    )

    expect(code).toMatch(
      `const _component_Qux = _${helperNameMap[RESOLVE_COMPONENT]}("Qux", true)\n`,
    )

    expect(code).toMatch(
      `const _directive_my_dir_0 = _${helperNameMap[RESOLVE_DIRECTIVE]}("my_dir_0")\n`,
    )
    expect(code).toMatch(
      `const _directive_my_dir_1 = _${helperNameMap[RESOLVE_DIRECTIVE]}("my_dir_1")\n`,
    )

    expect(code).toMatch('let _temp0, _temp1, _temp2')
  })

  test('hoists', () => {
    const root = createRoot({
      hoists: [
        createSimpleExpression(`hello`, false, locStub),
        createObjectExpression(
          [
            createObjectProperty(
              createSimpleExpression(`id`, true, locStub),
              createSimpleExpression(`foo`, true, locStub),
            ),
          ],
          locStub,
        ),
      ],
    })
    const { code } = generate(root)
    expect(code).toMatch(`const _hoisted_1 = hello`)
    expect(code).toMatch(`const _hoisted_2 = { id: "foo" }`)
    expect(code).toMatchSnapshot()
  })

  test('temps', () => {
    const root = createRoot({
      temps: 3,
    })
    const { code } = generate(root)
    expect(code).toMatch(`let _temp0, _temp1, _temp2`)
    expect(code).toMatchSnapshot()
  })

  test('static text', () => {
    const { code } = generate(
      createRoot({
        codegenNode: {
          type: NodeTypes.TEXT,
          content: 'hello',
          loc: locStub,
        },
      }),
    )

    expect(code).toMatch(`return "hello"`)
    expect(code).toMatchSnapshot()
  })

  test('interpolation', () => {
    const { code } = generate(
      createRoot({
        codegenNode: createInterpolation('hello', locStub),
      }),
    )

    expect(code).toMatch(`return _${helperNameMap[TO_DISPLAY_STRING]}(hello)`)
  })

  test('comment', () => {
    const { code } = generate(
      createRoot({
        codegenNode: {
          type: NodeTypes.COMMENT,
          content: 'foo',
          loc: locStub,
        },
      }),
    )

    expect(code).toMatch(`return _${helperNameMap[CREATE_COMMENT]}("foo")`)
  })

  test('compound expression', () => {
    const { code } = generate(
      createRoot({
        codegenNode: createCompoundExpression(
          [
            '_ctx.',
            createSimpleExpression('foo', false, locStub),
            ' + ',
            {
              type: NodeTypes.INTERPOLATION,
              loc: locStub,
              content: createSimpleExpression('bar', false, locStub),
            },
            createCompoundExpression([' + ', 'nested']),
          ],
          locStub,
        ),
      }),
    )

    expect(code).toMatch(
      `return _ctx.foo + _${helperNameMap[TO_DISPLAY_STRING]}(bar) + nested`,
    )
  })

  test('ifNode', () => {
    const { code } = generate(
      createRoot({
        codegenNode: {
          type: NodeTypes.IF,
          loc: locStub,
          // branches: [],
          codegenNode: createConditionalExpression(
            createSimpleExpression('foo', false),
            createSimpleExpression('bar', false),
            createSimpleExpression('baz', false),
          ),
        },
      }),
    )

    expect(code).toMatch(/return foo\s+\? bar\s+: baz/)
  })

  test('forNode', () => {
    const { code } = generate(
      createRoot({
        codegenNode: {
          type: NodeTypes.FOR,
          loc: locStub,
          source: createSimpleExpression('foo', false),
          valueAlias: undefined,
          keyAlias: undefined,
          codegenNode: {
            type: NodeTypes.VNODE_CALL,
            tag: FRAGMENT,
            isBlock: true,
            disableTracking: true,
            children: createCallExpression(RENDER_LIST),
          },
        },
      }),
    )

    expect(code).toMatch(`openBlock(true)`)
  })

  test('Element (callExpression + objectExpression + TemplateChildNode[])', () => {
    const { code } = generate(
      createRoot({
        codegenNode: createElementWithCodegen(
          `"div"`,
          createObjectExpression(
            [
              createObjectProperty(
                createSimpleExpression('id', true, locStub),
                createSimpleExpression('foo', true, locStub),
              ),
              createObjectProperty(
                createSimpleExpression(`prop`, false, locStub),
                createSimpleExpression(`bar`, false, locStub),
              ),
            ],
            locStub,
          ),
          [
            createElementWithCodegen(
              `"p"`,
              createObjectExpression(
                [
                  createObjectProperty(
                    // should quote the key!
                    createSimpleExpression(`some-key`, true, locStub),
                    createSimpleExpression(`foo`, true, locStub),
                  ),
                ],
                locStub,
              ),
            ),
          ],
          PatchFlags.FULL_PROPS,
        ),
      }),
    )

    expect(code).toMatch(
      `
    return _${helperNameMap[CREATE_ELEMENT_VNODE]}("div", {
      id: "foo",
      [prop]: bar
    }, [
      _${helperNameMap[CREATE_ELEMENT_VNODE]}("p", { "some-key": "foo" })
    ], ${genFlagText(PatchFlags.FULL_PROPS)})`,
    )

    expect(code).toMatchSnapshot()
  })
})
