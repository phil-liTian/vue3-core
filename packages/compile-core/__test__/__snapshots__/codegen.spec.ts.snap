// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: codegen > Element (callExpression + objectExpression + TemplateChildNode[]) 1`] = `
"return function render(_ctx, _cache) {
  with (_ctx) {
    return _createElementVNode("div", {
      id: "foo",
      [prop]: bar
    }, [
      _createElementVNode("p", { "some-key": "foo" })
    ], 16 /* FULL_PROPS */)
  }
}"
`;

exports[`compiler: codegen > function mode preamble 1`] = `
"const _Vue = Vue
return function render(_ctx, _cache) {
  with (_ctx) {
    const { createVNode: _createVNode, resolveDirective: _resolveDirective } = _Vue
return 
  }
}"
`;

exports[`compiler: codegen > function mode preamble w/ prefixIdentifiers: true 1`] = `
"const { createVNode: _createVNode, resolveDirective: _resolveDirective } = Vue
return function render(_ctx, _cache) {
  return 
}"
`;

exports[`compiler: codegen > hoists 1`] = `
"const _hoisted_1 = hello
const _hoisted_2 = { id: "foo" }
return function render(_ctx, _cache) {
  with (_ctx) {
    return 
  }
}"
`;

exports[`compiler: codegen > mode preamble 1`] = `
"import { createVNode as _createVNode, resolveDirective as _resolveDirective } from "vue"
function render(_ctx, _cache) {
  return 
}"
`;

exports[`compiler: codegen > static text 1`] = `
"return function render(_ctx, _cache) {
  with (_ctx) {
    return "hello"
  }
}"
`;

exports[`compiler: codegen > temps 1`] = `
"return function render(_ctx, _cache) {
  with (_ctx) {
    let _temp0, _temp1, _temp2return 
  }
}"
`;
